src/Razor/src/Microsoft.VisualStudio.Editor.Razor/DefaultDocumentDivergenceChecker.cs

```diff
@@ -0,0 +1,304 @@
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using DocumentSnapshot = Microsoft.CodeAnalysis.Razor.ProjectSystem.DocumentSnapshot;
+using SourceText = Microsoft.CodeAnalysis.Text.SourceText;
+using TextLineCollection = Microsoft.CodeAnalysis.Text.TextLineCollection;
+using TextSpan = Microsoft.CodeAnalysis.Text.TextSpan;
+
+namespace Microsoft.VisualStudio.Editor.Razor
+{
+    [Shared]
+    [Export(typeof(DocumentDivergenceChecker))]
+    internal class DefaultDocumentDivergenceChecker : DocumentDivergenceChecker
+    {
+        private static readonly RazorParserOptions ParserOptions = RazorParserOptions.Create(builder =>
+        {
+            builder.Directives.Add(FunctionsDirective.Directive);
+            builder.Directives.Add(ComponentCodeDirective.Directive);
+        });
+
+        // The goal of this method is to decide if the older and newer snapshots have diverged in a way that could result in
+        // impactful changes to other Razor files. Detecting divergence is an optimization which allows us to not always re-parse
+        // dependent files and therefore not not notify anyone else listening (Roslyn).
+        //
+        // Here's how we calculate "divergence":
+        //
+        // 1. Get the SyntaxTree for each document. If one has not been calculated yet, create a one-off syntax tree that only cares about
+        //    capturing Razor directives.
+        // 2. Extract @code and @functions directive blocks
+        // 3. Map those directive blocks back to the original source document and extract their inner content.
+        //    Aka @code { private int _foo; } => private int _foo;
+        // 4. Do a light-weight C# parse on the content of each captured @code/@functions content in order to build a C# SyntaxTree. The
+        //    SyntaxTree is not meant to be full-fidelity or error free, it's just meant to represent the key pieces of the content like
+        //    methods, fields, properties etc.
+        // 5. Extract all properties for the new and old documents.
+        // 6. Compare the old documents properties to the new documents properties, if they've changed structurally then we've diverged;
+        //    otherwise no divergence!
+        //
+        // At any point in this flow if we're unable to calculate one of the requirements such as the original Razor documents source text,
+        // assume divergence.
+        public override bool PossibleDivergence(DocumentSnapshot old, DocumentSnapshot @new)
+        {
+            if (old is null)
+            {
+                throw new ArgumentNullException(nameof(old));
+            }
+
+            if (@new is null)
+            {
+
+                throw new ArgumentNullException(nameof(@new));
+            }
+
+            if (!string.Equals(@new.FileKind, FileKinds.Component, StringComparison.OrdinalIgnoreCase))
+            {
+                // Component import or ordinary cshtml file.
+                return true;
```

Comment:
Why do we assume that a "regular" cshtml file will always have diverged?

Reply:
It's mostly that we don't have the logic to understand them appropriately so they fall under the "potentially diverged" category. In practice what happens is when our parser asks the project system for "dependent files", if it's a CSHTML file it'll almost always return 0 dependent files